requests_toolbelt/utils/dump.py

```diff
@@ -0,0 +1,166 @@
+"""This module provides functions for dumping information about responses."""
+import collections
+
+from requests import compat
+
+
+__all__ = ('dump_response',)
+_PrefixSettings = collections.namedtuple('PrefixSettings',
+                                         ['request', 'response'])
+HTTP_VERSIONS = {
+    9: b'0.9',
+    10: b'1.0',
+    11: b'1.1',
+}
+
+
+class PrefixSettings(_PrefixSettings):
+    def __init__(self, request, response):
+        request = _coerce_to_bytes(request)
+        response = _coerce_to_bytes(response)
+        super(PrefixSettings, self).__init__(request, response)
+
+
+def _format_header(name, value):
+    return (_coerce_to_bytes(name) + b': ' + _coerce_to_bytes(value) +
+            b'\r\n')
+
+
+def _dump_request_data(request, prefixes, bytearr):
+    prefix = prefixes.request
+    method = _coerce_to_bytes(request.method)
+    uri = compat.urlparse(request.url)
+
+    request_path = _coerce_to_bytes(uri.path)
+    if uri.query:
+        request_path + b'?' + _coerce_to_bytes(uri.query)
+
+    # <prefix><METHOD> <request-path> HTTP/1.1
+    bytearr.extend(prefix + method + b' ' + request_path + b' HTTP/1.1\r\n')
+
+    # <prefix>Host: <request-host> OR host header specified by user
+    headers = request.headers.copy()
+    host_header = _coerce_to_bytes(headers.pop('Host',
+                                   _coerce_to_bytes(uri.netloc)))
+    bytearr.extend(prefix + b'Host: ' + host_header + b'\r\n')
+
+    for name, value in sorted(headers.items()):
+        bytearr.extend(prefix + _format_header(name, value))
+
+    bytearr.extend(prefix + b'\r\n')
+    if request.body:
+        if isinstance(request.body, compat.basestring):
+            bytearr.extend(prefix + _coerce_to_bytes(request.body))
+        else:
+            # In the event that the body is a file-like object, let's not try
+            # to read everything into memory.
+            bytearr.extend('<< Request body is not a string-like type >>')
+    bytearr.extend(b'\r\n')
+
+
+def _dump_response_data(response, prefixes, bytearr):
+    prefix = prefixes.response
+    # Let's interact almost entirely with urllib3's response
+    raw = response.raw
+
+    # Let's convert the version int from httplib to bytes
+    version_str = HTTP_VERSIONS.get(raw.version, b'?')
+
+    # <prefix>HTTP/<version_str> <status_code> <reason>
+    bytearr.extend(prefix + b'HTTP/' + version_str + b' ' +
+                   str(raw.status).encode() + b' ' +
+                   _coerce_to_bytes(raw.reason) + b'\r\n')
+
+    headers = raw.headers
+    for name in sorted(headers.keys()):
+        for value in headers.getlist(name):
+            bytearr.extend(prefix + _format_header(name, value))
```

Comment:
I think reaching into httplib is a better idea here. With responses we can basically work out exactly what they looked like, so we should do that. This sorting of header names is really less than ideal.